pkg/kvstore/store/syncstore.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package store
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// SyncStore abstracts the operations allowing to synchronize key/value pairs
+// into a kvstore.
+type SyncStore interface {
+	// UpsertKey upserts a key/value pair into the kvstore.
+	UpsertKey(ctx context.Context, key Key) error
+
+	// DeleteLocalKey removes a key from the kvstore.
+	DeleteKey(ctx context.Context, key NamedKey) error
+
+	// Release frees all resources owned by the store, leaving all keys in the
+	// kvstore intact.
+	Release()
+}
+
+// SyncStoreBackend represents the subset kvstore.BackendOperations leveraged
+// by SyncStore implementations.
+type SyncStoreBackend interface {
+	// Update creates or updates a key.
+	Update(ctx context.Context, key string, value []byte, lease bool) error
+	// Delete deletes a key.
+	Delete(ctx context.Context, key string) error
+}
+
+// wqSyncStore implements the SyncStore interface leveraging a workqueue to
+// coalescence update/delete requests and handle retries in case of errors.
+type wqSyncStore struct {
+	backend   SyncStoreBackend
+	prefix    string
+	withLease bool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	workqueue workqueue.RateLimitingInterface
+	state     sync.Map /* map[string][]byte --- map[NamedKey.GetKeyName()]Key.Marhal() */
+
+	log *logrus.Entry
+	wg  sync.WaitGroup
+}
+
+// NewWorkqueueSyncStore returns a SyncStore instance which leverages a workqueue
+// to coalescence update/delete requests and handle retries in case of errors.
+func NewWorkqueueSyncStore(backend SyncStoreBackend, prefix string) SyncStore {
+	w := &wqSyncStore{
+		backend:   backend,
+		prefix:    prefix,
+		withLease: true,
+
+		workqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
```

Comment:
Yep, I forgot about that. Now it is configurable, and I also changed the `UpsertKey()` and `DeleteKey()` methods to perform a direct insertion in the workqueue (previously they incorrectly went through the rate limiter). Hence, the limiter is only used in case of requeues, since kvstore operations are already rate-limited by a separated rate limiter.